### ISSUE TITLE
[RFC] Integrate ppx_view

### ISSUE THREAD

**Issue by diml** [ocaml-ppx/ppxlib#47]
This ticket describe how to integrate [ppx_view](https://github.com/janestreet/ppx_view) in ppxlib. This is the last major work to do before we release the 1.0.0 version of ppxlib. These two projects could stay separate, however given that using view pattern will improve a lot the ppx compatibility story, I believe that it is worth merging them to promote the use of view patterns.

The rest of this PR explains what needs to be done.

## Structure of ppx_view

`ppx_view` is composed of:

- `ppx_view.lib`, a library defining the type of view patterns + some combinators. It is used as the runtime library of the ppx rewriter
- `ppx_view`, a ppx rewriter that interprets `match%view ...` expressions and produces code using the `ppx_view.lib` library
- `ppx_view.ast`, a library providing view patterns for the OCaml AST. Its API is the dual of the `Ast_helper` module from the compiler libraries

A lot of the code of `ppx_view.ast` is automatically generated from parsetree.mli.

`ppx_view.lib` and `ppx_view.ast` can be integrated in `ppxlib` as it. Most of the work will be on integrating `ppx_view.ast` so that the overall API of ppxlib stays consistent and is ready for long term backward compatibility.

## Ast_pattern and ppx_view.ast

`ppxlib` currently has a `Ppxlib.Ast_pattern` module that serve a purpose similar to the one of `ppx_view.ast`. However, the API of `Ppxlib.Ast_pattern` follows the naming convention of `Ppxlib.Ast_builder` rather than `Ast_helper`.

It doesn't make sense to maintain both `Ppxlib.Ast_pattern` and the original `ppx_view.ast`. View patterns are better than what `Ppxlib.Ast_pattern` currently provides, so we should migrate `Ast_pattern` to use plain view patterns instead. However, to make it easier  to upgrade code currently using `Ppxlib.Ast_pattern`, we should keep the current naming conventions of `Ast_pattern`.

Most of the combinators of `Ast_pattern` are generated, so it should be easy to adapt the generator to generate view patterns.

## Backward compatibility story

As the OCaml AST evolves, we will update the API of `Ast_builder` and `Ast_pattern` in backward compatible ways. For instance by adding new `foo_v409` functions, or by clever use of optional arguments. One consequence is that we can no longer generate the interface of `Ast_builder` and `Ast_pattern`. Instead, we should just infer the mli of `Ast_pattern_generated` and include it in `Ast_pattern` and do the same for `Ast_builder`. The implementation can still be generated, and we will overload a few functions in `Ast_pattern` to make sure to preserve backward compatibility when the signature of the generated functions change.

While this will work fine, over time the API will become quite ugly and we will want to clean it up. For this reason, we should version the whole API of `Ast_builder` and `Ast_pattern`. I.e. `Ast_builder` will become `Ast_builder.V1` and `Ast_pattern` will become `Ast_pattern.V1`. When the API will become too messy, we will mint `V2` versions.

Note: to make things simpler, we can get rid of `Ast_builder.Make`. This function is almost never used and it is more code to maintain.

**Comment by trefis** [comment]
I'd like to have a look at this before it gets merged if you don't mind.
I'll approve / comment when I'm done.

**Comment by diml** [comment]
Sure

**Comment by rgrinberg** [comment]
Possibly a naive question, but why would do we need to support both builder & helper style for patterns & constructors? Seems like pattern/builder provide nicer conventions, can't we just use those exclusively?

**Comment by xclerc** [comment]
If memory serves, it was basically to explore the design space.
I implemented toy versions of various ppx rewriters to see how
it would look like; feel free to remove the cruft.

**Comment by diml** [comment]
The idea is indeed to support only pattern/builder style. However, the current `Ast_pattern` should be re-implemented on top of the `View` module.

**Comment by diml** [comment]
I was thinking about this again. If we simply upstream the `View` and `Ast_viewer` modules of `ppx_view` into the compiler directly, then we would get the best possible world: all the ppx ecosystem could use the current AST without any conversion and without risk of breaking with new compiler versions, and the for compiler updating `Ast_viewer` when changing the parsetree would be no harder than updating `Ast_helper`.

**Comment by diml** [comment]
I'm closing this for now. This plan is a step forward but doesn't go far enough. The new plan is to add first a minimal and stable API for ppx rewriters in the compiler, then we will be able to refactor ppxlib and other ppx libraries on top of this new library. Once this is done, changes to the compiler libraries should cause no breakage in the ppx ecosystem. In particular, it will become possible to test development compilers against released opam packages.

I'm currently writing a proposal for the stable API.
